### PR TITLE
fixes requires for pygtk

### PR DESCRIPTION
**Pull request by lee218llnl**
These are some changes that I needed to get pygtk installed and running properly.

### CODE REVIEW

File: var/spack/repos/builtin/packages/pango/package.py [hunk 1]
```diff
@@ -43,7 +43,6 @@ class Pango(AutotoolsPackage):
 
     depends_on("pkg-config", type="build")
     depends_on("harfbuzz")
-    depends_on("cairo")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pango/package.py]: Why was this change necessary? With this change, if I run:
```
$ spack spec pango ^cario@1.2.3
```
it will complain that "pango does not depend on cairo". This is a bug in the concretization algorithm. This line was the current workaround. Same for `gtkplus`.

Comment (lee218llnl) [inline var/spack/repos/builtin/packages/pango/package.py]: @adamjstewart I can put it back if you think it is necessary, but I just ran

spack spec pango^cairo@1.14.8

and it returned a concretized spec for me.

Comment (lee218llnl) [inline var/spack/repos/builtin/packages/pango/package.py]: @adamjstewart and in fact without these changes (i.e. if you just checkout the develop branch), when I run:

spack spec py-pygtk

I get: 

Package cairo requires variant ~X, but spec asked for +X

But then if I run:

spack spec py-pygtk^cairo~X

I get:

Error: Invalid spec: 'cairo~X'. Package cairo requires variant +X, but spec asked for ~X

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pango/package.py]: What does your `packages.yaml` look like? With no `packages.yaml`, it should concretize correctly. I made sure that the `X` variant for every package was disabled by default, and the `mpi` variant is enabled for every package by default. If you change these defaults in your `packages.yaml` or on the command line, you're screwed unless you change the dependencies too. Hence the broken concretizer.

File: var/spack/repos/builtin/packages/py-pygobject/package.py [hunk 2]
```diff
@@ -31,18 +31,25 @@ class PyPygobject(AutotoolsPackage):
...
     depends_on('glib')
     depends_on('py-py2cairo')
     depends_on('gobject-introspection')
 
-    patch('pygobject-2.28.6-introspection-1.patch')
+    patch('pygobject-2.28.6-introspection-1.patch', when='@2.28.3:2.28.6')
+
+    # patch from https://raw.githubusercontent.com/NixOS/nixpkgs/master/
+    #                   pkgs/development/python-modules/pygobject/
+    #                   pygobject-2.28.6-gio-types-2.32.patch
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/py-pygobject/package.py]: Can you put this URL on one line so we can cut-n-paste it easily?  @adamjstewart does this need help with flake8?

File: var/spack/repos/builtin/packages/py-pygtk/package.py [hunk 3]
```diff
@@ -36,8 +36,8 @@ class PyPygtk(AutotoolsPackage):
     depends_on("libffi")
     depends_on('cairo')
     depends_on('glib')
-    depends_on('gtkplus')
-    depends_on('py-pygobject')
+    depends_on('gtkplus@2.24:2.99')
+    depends_on('py-pygobject@2.28:2.99')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/py-pygtk/package.py]: This looks correct... that py-gtk is GTK2-only.  Can you add a comment (here and in the docstring at the top) pointing users to `py-pygobject` if they want to do Python-based GTK3?

File: var/spack/repos/builtin/packages/py-pygobject/package.py [hunk 4]
```diff
@@ -31,18 +31,25 @@ class PyPygobject(AutotoolsPackage):
...
     homepage = "https://pypi.python.org/pypi/pygobject"
 
-    # FIXME: This URL is no longer available for download from PyPi
-    url      = "https://pypi.python.org/packages/6d/15/97c8b5ccca2be14cf59a2f79e15e3a82a1c3408a6b76b4107689a8b94846/pygobject-2.28.3.tar.bz2"
+    url      = "http://ftp.gnome.org/pub/GNOME/sources/pygobject/2.28/pygobject-2.28.6.tar.bz2"
 
+    version('2.28.6', 'a43d783228dd32899e6908352b8308f3')
     version('2.28.3', 'aa64900b274c4661a5c32e52922977f9')
 
     extends('python')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/py-pygobject/package.py]: Do you know whether this works with Python2, Python3 or both?

File: var/spack/repos/builtin/packages/pango/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pango/package.py]: I would say yes, if it had any benefit whatsoever. But if it doesn't, it makes it harder to test specific versions of dependencies.

File: var/spack/repos/builtin/packages/py-pygobject/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pygobject/package.py]: Nope, any line containing `(file|https?)://` gets excluded from the flake8 line length checks.

**Comment by citibeth** [comment]
OK... Can we add a comment to the line, so we know it can be removed once
we have a better concretizer?

**Comment by lee218llnl** [comment]
@adamjstewart and @citibeth thanks for all the feedback, I believe this latest commit addresses all of your comments.

File: var/spack/repos/builtin/packages/py-py2cairo/package.py [hunk 5]
```diff
@@ -35,7 +35,7 @@ class PyPy2cairo(Package):
 
     extends('python')
 
-    depends_on('cairo+X')
+    depends_on('cairo')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-py2cairo/package.py]: I'm actually the one who added `+X` to this dependency. When I was working on #2778, I noticed that this package did not build for me because it was looking for X11 dependencies. Adding `+X` solved the problem for me.

Comment (citibeth) [inline var/spack/repos/builtin/packages/py-py2cairo/package.py]: So... one person says that `cairo+X` is required to build, another says that the `+X` breaks things.  Can anyone explain at a deeper level what the core problem is?

Comment (lee218llnl) [inline var/spack/repos/builtin/packages/py-py2cairo/package.py]: OK, I can put this back and just have py-pygtk depend on gtkplus+X. The reason I made the change was that py-pygtk would fail with Package cairo requires variant ~X, but spec asked for +X.

**Comment by citibeth** [comment]
On the `cairo+X` issue, please review #2575.  As @adamjstewart pointed out, `~` should probably never be part of a `depends_on()` declaration.  I forget the reason why we have it anyway.  But it clearly causes cascading problems, and would be great if we can get rid of it.

File: var/spack/repos/builtin/packages/py-py2cairo/package.py (hunk 5 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-py2cairo/package.py]: Yeah, the concretizer is pretty broken. Ping @tgamblin 😆

**Comment by lee218llnl** [comment]
@tgamblin I think this is ready to be merged.
